File: java/components/StrataPersistence/src/strata1/persistence/inmemoryrepository/InMemoryFinder.java
// ##########################################################################
// # File Name:	InMemoryNamedQuery.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.persistence.inmemoryrepository;


import strata1.persistence.repository.AbstractFinder;
import strata1.persistence.repository.IFinder;
import strata1.persistence.repository.NotUniqueException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/****************************************************************************
 * 
 * @param <K> key type
 * @param <T> entity type
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class InMemoryFinder<T> 
    extends    AbstractFinder<T>
	implements IFinder<T>
{
	private InMemoryRepositoryContext itsContext;
	private Class<T>                  itsType;
	private IPredicate<T>			  itsPredicate;
	private List<T>				      itsObjects;
	private List<T>				      itsResults;
	private Iterator<T>				  itsIterator;
		
	/************************************************************************
	 * Creates a new InMemoryNamedQuery. 
	 * @param type TODO
	 *
	 */
	public 
	InMemoryFinder(
	    InMemoryRepositoryContext context,
		String                    name,
		Class<T>                  type, 
		IPredicate<T>			  predicate)
	{
		super(type.getCanonicalName() + "." + name);
		itsContext       = context;
		itsType          = type;
		itsPredicate	 = predicate;
		itsObjects       = itsContext.getEntitiesByType( itsType );
		itsResults       = null;
		itsIterator      = null;
		
		itsContext.insertFinder(type,this);
	}

	/************************************************************************
	 * Creates a new InMemoryNamedQuery. 
	 *
	 */
	public 
	InMemoryFinder(InMemoryFinder<T> other)
	{
		super(other);
		itsContext       = other.itsContext;
		itsType          = other.itsType;
		itsPredicate	 = other.itsPredicate;
		itsObjects       = itsContext.getEntitiesByType( itsType );
		itsResults       = null;
		itsIterator      = null;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public InMemoryFinder<T> 
	copy()
	{
	    return new InMemoryFinder<T>( this );
	}

	/************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    clear()
    {
        itsObjects = itsContext.getEntitiesByType( itsType );
        itsResults = null;
        itsIterator = null;
    }

    /************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public T 
	getUnique() 
	    throws NotUniqueException
	{
		execute();
		
		if ( itsResults.size() > 1 )
		    throw new NotUniqueException("result is not unique");
		
		if ( itsResults.isEmpty() )
		    return null;
		
		return itsResults.get( 0 );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public Collection<T> 
	getAll()
	{
		execute();
		return itsResults;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public T 
	getNext()
	{
		execute();
		return itsIterator.next();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	public boolean 
	hasNext()
	{
		execute();
		return itsIterator.hasNext();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
    public boolean 
    hasUnique() 
        throws NotUniqueException
    {
	    execute();
	    return itsResults.size() == 1;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    hasAny()
    {
        execute();
        return !itsResults.isEmpty();
    }

    /************************************************************************
	 *  
	 *
	 */
	protected void
	execute()
	{
		if ( itsResults != null )
			return;
		
		itsResults  = getResults( getInputs().getNamedInputs(),itsObjects );
		itsIterator = itsResults.iterator();
	}
	
	/************************************************************************
	 *  
	 *
	 * @param inputs
	 * @param objects
	 * @return
	 */
	protected List<T> 
	getResults(Map<String,Object> inputs,List<T> objects)
	{
		List<T> output = new ArrayList<T>();
		
		for (T entry : objects)
			if ( itsPredicate.evaluate( entry,inputs ) )
				output.add( entry );
		
		return output;
	}
}


// ##########################################################################

File: legacy/csharp/components/StrataPersistence/InMemoryRepository/InMemoryFinder.cs
//  ##########################################################################
//  # File Name: InMemoryFinder.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################
 
using System;
using System.Collections.Generic;
using System.Linq;
using Strata.Common.Utility;
using Strata.Persistence.Repository;

namespace Strata.Persistence.InMemoryRepository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    /// <author>JFL</author>
    /// <conventions>$conventionspath$</conventions>
    ///  
    public 
    class InMemoryFinder<T>:
        AbstractFinder<T>
        where T:class
    {	
        private InMemoryRepositoryContext itsContext;
	    private Type                      itsType;
	    private IPredicate<T>			  itsPredicate;
	    private IList<T>				  result;
	    private IEnumerator<T>			  current;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
        public 
        InMemoryFinder(
            InMemoryRepositoryContext cntx,
            String                    name,
            IPredicate<T>             pred):
            base(name)
        {
            itsContext   = cntx;
		    itsType      = typeof(T);
		    itsPredicate = pred;
		    result   = null;
		    current  = null;
		
		    itsContext.InsertFinder<T>(this);
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
        public   
        InMemoryFinder(InMemoryFinder<T> other):
            base( other )
        {
		    itsContext   = other.itsContext;
		    itsType      = other.itsType;
		    itsPredicate = other.itsPredicate;
		    result       = null;
		    current      = null;
        } 

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override ICopyable
        MakeCopy()
        {
            return new InMemoryFinder<T>( this );            
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.GetAll()"/>
        /// </summary>
        /// 
        public override void 
        Clear()
        {
            ClearInputs();
            result  = null;
            current = null;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.HasNext()"/>
        /// </summary>
        /// 
	    public override void
	    Execute()
	    {
		    if ( result != null )
			    return;
		
		    result  = GetResults( 
                GetInputs().GetNamedInputs(),
                itsContext.GetEntitiesByType<T>() );
		    current = result.GetEnumerator();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.GetAll()"/>
        /// </summary>
        /// 
        public override ICollection<T> 
        GetAll()
        {
		    Execute();
		    return result;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.GetUnique()"/>
        /// </summary>
        /// 
        public override T 
        GetUnique()
        {
		    Execute();

            return result.SingleOrDefault();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.GetNext()"/>
        /// </summary>
        /// 
        public override T 
        GetNext()
        {
		    Execute();

            try
            {
                T next = current.Current;
                
                current.MoveNext();
                return next;
            }
            catch (InvalidOperationException)
            {
                return null;
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.HasUnique()"/>
        /// </summary>
        /// 
        public override bool 
        HasUnique()
        {
            Execute();
            return result.Count == 1;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.HasAny()"/>
        /// </summary>
        /// 
        public override bool 
        HasAny()
        {
            Execute();
            return result.Count > 0;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.HasNext()"/>
        /// </summary>
        /// 
        public override bool 
        HasNext()
        {
		    Execute();

            try
            {
                T next = current.Current;
                
                return true;
            }
            catch (InvalidOperationException)
            {
                return false;
            }
        }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IFinder{T}.HasNext()"/>
        /// </summary>
        /// 
	    protected IList<T> 
	    GetResults(IDictionary<String,Object> inputs,IList<T> objects)
	    {
		    IList<T> output = new List<T>();
		
		    foreach (T entry in objects)
			    if ( itsPredicate.Evaluate( entry,inputs ) )
				    output.Add( entry );
		
		    return output;
	    }

    }
}

//  ##########################################################################
